cmd/libsnap-confine-private/tool.c

```diff
@@ -0,0 +1,183 @@
+}
+
+void sc_call_snap_update_ns(int snap_update_ns_fd, const char *snap_name,
+			    struct sc_apparmor *apparmor)
+{
+	char *snap_name_copy SC_CLEANUP(sc_cleanup_string) = NULL;
+	snap_name_copy = sc_strdup(snap_name);
+	char *argv[] =
+	    { "snap-update-ns", "--from-snap-confine", snap_name_copy, NULL };
+	char *envp[] = { "SNAPD_DEBUG=x", NULL };
+	char profile[PATH_MAX] = { 0 };
+	sc_must_snprintf(profile, sizeof profile, "snap-update-ns.%s",
+			 snap_name);
+	sc_call_snapd_tool_with_apparmor(snap_update_ns_fd, "snap-update-ns",
+					 apparmor, profile, argv, envp);
+}
+
+int sc_open_snap_discard_ns(void)
+{
+	return sc_open_snapd_tool("snap-discard-ns");
+}
+
+void sc_call_snap_discard_ns(int snap_discard_ns_fd, const char *snap_name)
+{
+	char *snap_name_copy SC_CLEANUP(sc_cleanup_string) = NULL;
+	snap_name_copy = sc_strdup(snap_name);
+	char *argv[] =
+	    { "snap-discard-ns", "--from-snap-confine", snap_name_copy, NULL };
+	char *envp[] = { "SNAPD_DEBUG=x", NULL };
+	sc_call_snapd_tool(snap_discard_ns_fd, "snap-discard-ns", argv, envp);
+}
+
+static int sc_open_snapd_tool(const char *tool_name)
+{
+	// +1 is for the case where the link is exactly PATH_MAX long but we also
+	// want to store the terminating '\0'. The readlink system call doesn't add
+	// terminating null, but our initialization of buf handles this for us.
+	char buf[PATH_MAX + 1] = { 0 };
+	if (readlink("/proc/self/exe", buf, sizeof buf) < 0) {
+		die("cannot readlink /proc/self/exe");
+	}
+	if (buf[0] != '/') {	// this shouldn't happen, but make sure have absolute path
+		die("readlink /proc/self/exe returned relative path");
+	}
+	char *dir_name = dirname(buf);
+	int dir_fd SC_CLEANUP(sc_cleanup_close) = 1;
+	dir_fd = open(dir_name, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+	if (dir_fd < 0) {
+		die("cannot open path %s", dir_name);
+	}
+	int tool_fd = -1;
+	tool_fd = openat(dir_fd, tool_name, O_PATH | O_NOFOLLOW | O_CLOEXEC);
+	if (tool_fd < 0) {
+		die("cannot open path %s/%s", dir_name, tool_name);
+	}
+	debug("opened %s executable as file descriptor %d", tool_name, tool_fd);
+	return tool_fd;
+}
+
+static void sc_call_snapd_tool(int tool_fd, const char *tool_name, char **argv,
+			       char **envp)
+{
+	sc_call_snapd_tool_with_apparmor(tool_fd, tool_name, NULL, NULL, argv,
+					 envp);
+}
+
+static void sc_call_snapd_tool_with_apparmor(int tool_fd, const char *tool_name,
+					     struct sc_apparmor *apparmor,
+					     const char *aa_profile,
+					     char **argv, char **envp)
+{
+	debug("calling snapd tool %s", tool_name);
+	pid_t child = fork();
+	if (child < 0) {
+		die("cannot fork to run snapd tool %s", tool_name);
+	}
+	if (child == 0) {
+		/* If the caller provided template environment entry for SNAPD_DEBUG
+		 * then expand it to the actual value. */
+		for (char **env = envp;
+		     /* Mama mia, that's a spicy meatball. */
+		     env != NULL && *env != NULL && **env != '\0'; env++) {
+			if (sc_streq(*env, "SNAPD_DEBUG=x")) {
```

Comment:
Ah, I think I get it now, mind adding an extra comment above whe `SNAPD_DEBUG=x` is set?

Reply:
Not at all, I'll do that straight away :)